apps/cms/src/pages/api/sync-products-variants.ts

```diff
@@ -0,0 +1,176 @@
+import { providersSchemaSet } from "../../lib/cms/config/providers";
+import { cmsProviders, CMSProvider } from "../../lib/cms/providers";
+import { logger as pinoLogger } from "../../lib/logger";
+import { createClient } from "../../lib/graphql";
+import { createSettingsManager } from "../../lib/metadata";
+import {
+  executeBatchMetadataUpdate,
+  MetadataRecord,
+} from "../../lib/cms/client/metadata-execution";
+import { CmsBatchOperations } from "../../lib/cms/types";
+
+export interface SyncProductsVariantsApiPayload {
+  channelSlug: string;
+  providerInstanceId: string;
+  productsVariants: WebhookProductVariantFragment[];
+  operation: "ADD" | "DELETE";
+}
+
+export interface SyncProductsVariantsApiResponse {
+  success: boolean;
+  data?: {
+    createdCMSIds: MetadataRecord[];
+    deletedCMSIds: MetadataRecord[];
+  };
+  error?: string;
+}
+
+const handler: NextProtectedApiHandler = async (
+  req: NextApiRequest,
+  res: NextApiResponse<SyncProductsVariantsApiResponse>,
+  context
+) => {
+  const { authData } = context;
+
+  const logger = pinoLogger.child({
+    endpoint: "sync-products-variants",
+  });
+  logger.debug("Called endpoint sync-products-variants");
+
+  const client = createClient(authData.saleorApiUrl, async () => ({
+    token: authData.token,
+  }));
+
+  if (req.method !== "POST") {
+    return res.status(405).json({
+      success: false,
+    });
+  }
+
+  // todo: change to zod validation
+  const { providerInstanceId, productsVariants, operation } =
+    req.body as SyncProductsVariantsApiPayload;
+
+  if (!providerInstanceId) {
+    return res.status(400).json({
+      success: false,
+      error: "The provider instance id is missing.",
+    });
+  }
+
+  if (!productsVariants || productsVariants?.length === 0) {
+    return res.status(400).json({
+      success: false,
+      error: "The products variants are missing.",
+    });
+  }
+
+  if (!operation || (operation !== "ADD" && operation !== "DELETE")) {
+    return res.status(400).json({
+      success: false,
+      error: "The operation is missing or invalid. Allowed operations: ADD, DELETE.",
+    });
+  }
+  const operationType: keyof CmsBatchOperations =
+    operation === "ADD" ? "createBatchProducts" : "deleteBatchProducts";
+
+  const settingsManager = createSettingsManager(client);
+  const providerInstancesSettingsParsed = await getProviderInstancesSettings(settingsManager);
+  const providerInstanceSettings = providerInstancesSettingsParsed[providerInstanceId];
+  const channelsSettingsParsed = await getChannelsSettings(settingsManager);
+
+  const provider = cmsProviders[
+    providerInstanceSettings.providerName as CMSProvider
+  ] as (typeof cmsProviders)[keyof typeof cmsProviders];
+  const validation =
+    providersSchemaSet[providerInstanceSettings.providerName as CMSProvider].safeParse(
+      providerInstanceSettings
+    );
+
+  logger.debug("The provider instance settings provider.");
+  logger.debug({
+    provider,
+  });
```

Comment:
```suggestion
  logger.debug({ provider }, "The provider instance settings provider.");
```

Just throwing it out there in case you don't know about it. If you prefer to have it separately, that's okay.

Reply:
I don't see in console the second argument

Reply:
@orzechdev you have to follow the printf formatting style for multiple arguments https://getpino.io/#/docs/api?id=message-string

Reply:
Ok, I was supplying the object as the second argument, I thought it didn't matter. Now it works, thank you.